src/llvm-remove-addrspaces.cpp

```diff
@@ -0,0 +1,461 @@
+                }
+                else
+                    DstTy = StructType::get(
+                            Ty->getContext(), NewElTys, Ty->isPacked());
+            }
+        }
+        else if (auto Ty = dyn_cast<ArrayType>(SrcTy))
+            DstTy = ArrayType::get(
+                    remapType(Ty->getElementType()), Ty->getNumElements());
+        else if (auto Ty = dyn_cast<VectorType>(SrcTy))
+            DstTy = VectorType::get(
+                    remapType(Ty->getElementType()),
+                    Ty->getNumElements(),
+                    Ty->isScalable());
+
+        if (DstTy != SrcTy)
+            LLVM_DEBUG(
+                    dbgs() << "Remapping type:\n"
+                           << "  from " << *SrcTy << "\n"
+                           << "  to   " << *DstTy << "\n");
+
+        MappedTypes[SrcTy] = DstTy;
+        return DstTy;
+    }
+
+private:
+    static DenseMap<Type *, Type *> MappedTypes;
+};
+
+DenseMap<Type *, Type *> AddrspaceRemoveTypeRemapper::MappedTypes;
+
+class AddrspaceRemoveValueMaterializer : public ValueMaterializer {
+    ValueToValueMapTy &VM;
+    RemapFlags Flags;
+    ValueMapTypeRemapper *TypeMapper = nullptr;
+
+public:
+    AddrspaceRemoveValueMaterializer(
+            ValueToValueMapTy &VM,
+            RemapFlags Flags = RF_None,
+            ValueMapTypeRemapper *TypeMapper = nullptr)
+        : VM(VM), Flags(Flags), TypeMapper(TypeMapper)
+    {
+    }
+
+    Value *materialize(Value *SrcV)
+    {
+        Value *DstV = nullptr;
+        if (auto CE = dyn_cast<ConstantExpr>(SrcV)) {
+            Type *Ty = remapType(CE->getType());
+            if (CE->getOpcode() == Instruction::AddrSpaceCast) {
+                // peek through addrspacecasts if their address spaces match
+                // (like RemoveNoopAddrSpaceCasts, but for const exprs)
+                Constant *Src = mapConstant(CE->getOperand(0));
+                if (Src->getType()->getPointerAddressSpace() ==
+                    Ty->getPointerAddressSpace())
+                    DstV = Src;
+            }
+            else {
+                // recreate other const exprs with their operands remapped
+                SmallVector<Constant *, 4> Ops;
+                for (unsigned Index = 0; Index < CE->getNumOperands();
+                     ++Index) {
+                    Constant *Op = CE->getOperand(Index);
+                    Constant *NewOp = mapConstant(Op);
+                    Ops.push_back(NewOp ? cast<Constant>(NewOp) : Op);
+                }
+
+                if (CE->getOpcode() == Instruction::GetElementPtr) {
+                    // GEP const exprs need to know the type of the source.
+                    // asserts remapType(typeof arg0) == typeof mapValue(arg0).
+                    Constant *Src = CE->getOperand(0);
+                    Type *SrcTy = remapType(
+                            cast<PointerType>(Src->getType()->getScalarType())
+                                    ->getElementType());
```

Comment:
```suggestion
                     Type *SrcTy = remapType(cast<GetElementPtrConstantExpr>(CE)
                             ->getSourceElementType());
```

Reply:
I don't think we can use GetElementPtrConstantExpr, it looks like a private class?